### PR TITLE
feat(rust): Implement Enhanced Query Parameter Builder with Advanced Filtering Support

### LINKED ISSUE #8847: Rust SDK: Implement Complex Query Parameter Builder for Advanced API Filtering
### Which Fern component?

SDK Generator

### How important is this?

P0 - Critical (Blocking work)

### What's the feature?

**Description**:
Implement advanced query parameter building to support complex filtering patterns found in production APIs like Foxglove.

**Current State**:
```rust
// Current - VERY LIMITED
pub async fn list_devices(limit: Option<String>) -> Result<Vec<Device>, Error>
// Only supports: ?limit=50
```

**Required Implementation**:
```rust
// Target - COMPREHENSIVE
pub async fn list_devices(params: DeviceQueryParams) -> Result<Vec<Device>, Error>

#[derive(Debug, Clone, Default)]
pub struct DeviceQueryParams {
    pub query: Option<String>,  // For complex filtering
    pub sort_by: Option<DeviceSortBy>,
    pub sort_order: Option<SortOrder>,
    pub limit: Option<u32>,
    pub offset: Option<u32>,
    pub additional_params: HashMap<String, String>,
}

// Support complex queries like:
// ?query=properties.locationId:lab1,lab2 properties.status:active name:robot
// ?sortBy=device.name&sortOrder=asc&limit=100&offset=200
```

**Key Requirements from Foxglove API**:
1. **Complex Filter Syntax**: `properties.key:value1,value2` (comma-separated OR)
2. **Nested Queries**: `properties.key:*` (wildcard), `*:value` (any field)
3. **Multi-qualifiers**: `key1:value1 key2:value2` (space-separated AND)
4. **Special Characters**: `"value with spaces"` (quoted values)
5. **Proper Encoding**: URL-safe encoding of special characters

**Acceptance Criteria**:
- [ ] Supports Foxglove-style complex query syntax: `properties.key:value1,value2`
- [ ] Handles quoted values with spaces: `name:"My Device"`
- [ ] Implements AND logic for space-separated queries
- [ ] Proper URL encoding of special characters
- [ ] Type-safe sort field enums generated per endpoint
- [ ] Integration tests with real Foxglove API patterns

### Any alternatives?

_No response_

### Are you interested in contributing this feature?

Yes

### PR DESCRIPTION
**Pull request by iamnamananand996**
closes: https://github.com/fern-api/fern/issues/8847

## Description

Implemented a query parameter builder system for Rust SDK generation that supports advanced API filtering patterns. The feature introduces a new `QueryParameterBuilder` utility for structured query parsing and enhanced query parameter handling, along with improved code organization and type safety throughout the SubClientGenerator.

This enhancement enables Rust SDKs to handle complex query scenarios including structured query strings (e.g., `"key:value key2:value1,value2"`), URL encoding, and fallback mechanisms for backwards compatibility.

## Changes Made

- **Added QueryParameterBuilder utility** - New Rust module for advanced query parameter handling with URL encoding, structured query parsing, and error handling
- **Enhanced SubClientGenerator architecture** - Completely reorganized code structure with logical groupings: Client Generation → HTTP Methods → Parameter Extraction → Query Building → Type Utilities → Pagination Support
- **Implemented enhanced query parameter detection** - Automatic detection of complex endpoints requiring advanced query building (endpoints with >5 params, "query"/"filter" params, or sort parameters)
- **Removed code duplication** - Consolidated duplicated query parameter building logic into shared `buildQueryParameterStatements()` method

## Testing

- **Unit tests added/updated** -Updated test coverage for QueryParameterBuilder with URL encoding, structured queries, and edge cases
- **Manual testing completed** - Successfully tested query-parameters fixture

### CODE REVIEW

File: seed/rust-sdk/nullable/src/client/nullable.rs [hunk 1]
```diff
@@ -12,26 +12,26 @@ impl NullableClient {
...
     }
 
-    pub async fn get_users(&self, usernames: Option<Option<String>>, avatar: Option<Option<String>>, activated: Option<Option<bool>>, tags: Option<Option<Option<String>>>, extra: Option<Option<Option<bool>>>, options: Option<RequestOptions>) -> Result<Vec<User>, ClientError> {
+    pub async fn get_users(&self, usernames: Option<String>, avatar: Option<String>, activated: Option<bool>, tags: Option<Option<String>>, extra: Option<Option<bool>>, options: Option<RequestOptions>) -> Result<Vec<User>, ClientError> {
         self.http_client.execute_request(
             Method::GET,
             "/users",
             None,
             {
             let mut query_params = Vec::new();
```
Comment (musicpulpite) [inline seed/rust-sdk/nullable/src/client/nullable.rs:21]: Why does this seed case not use the new `query_builder` pattern?

Comment (musicpulpite) [reply seed/rust-sdk/nullable/src/client/nullable.rs:21]: NVM presumably because you have the `shouldUseEnhancedQueryBuilder` check?

**Review by musicpulpite — APPROVED** [review]
Excellent work! Neither of the test failures are concerning here

File: seed/rust-sdk/nullable/src/client/nullable.rs (hunk 1 above)
Comment (iamnamananand996) [reply seed/rust-sdk/nullable/src/client/nullable.rs:21]: Hi @musicpulpite,

Yes correct, maybe in the next revision, we can enhance these conditions more to cover more case.

```typescript
private shouldUseEnhancedQueryBuilder(endpoint: HttpEndpoint): boolean {
        const queryParams = endpoint.queryParameters;

        // Use enhanced builder if:
        // 1. There's a parameter named "query" (structured query string)
        // 2. There are many query parameters (>5) suggesting complex filtering
        // 3. There's a mix of sort parameters (sortBy, sortOrder) indicating advanced querying
        return (
            queryParams.some(
                (param) =>
                    param.name.wireValue === "query" || // Structured query parameter
                    param.name.wireValue === "filter" || // Generic filter parameter
                    param.name.wireValue.includes("sort") // Sort-related parameters
            ) || queryParams.length > 5
        ); // Many parameters suggest complex usage
  }
```

**Comment by iamnamananand996** [comment]
> Excellent work! Neither of the test failures are concerning here

Hi @musicpulpite, thank you for review the PR.

The failed tests will always fail for me, as an outside contributor, those 2 tests will require organization member access, this is a known issue. 😄

**Comment by musicpulpite** [comment]
> > Excellent work! Neither of the test failures are concerning here
> 
> Hi @musicpulpite, thank you for review the PR.
> 
> The failed tests will always fail for me, as an outside contributor, those 2 tests will require organization member access, this is a known issue. 😄

Yep exactly haha. Makes sense

**Comment by musicpulpite** [comment]
Hey @iamnamananand996 it looks like you created all of these PR's off of a shared base branch or something so please resolve the merge conflicts and then I will be happy to merge on your behalf!

**Comment by iamnamananand996** [comment]
closing this also because of https://github.com/fern-api/fern/pull/8877#issuecomment-3256153986
